### ISSUE TITLE
Getting data out of an Error, easier classifying unknown values, "Type" module thoughts

### ISSUE THREAD

**Issue by jmagaram** [rescript-association/rescript-core#102]
I'm trying to do something simple. I caught a Firebase exception and want to access the `code` property that is a string. Firebase errors have code properties with contents like "auth/email-password-invalid". How can I do that? Here is one awkward way. **Is there a simpler way?**
```
@get_index external getErrorProperty: (Error.t, string) => option<'a> = ""

let toString1 = i =>
  switch i->Type.Classify.classify {
  | String(s) => Some(s)
  | _ => None
  }

let printError = (err: Error.t) =>
  err
  ->getErrorProperty("code")
  ->toString1
  ->Option.getWithDefault("No code found in the exception")
  ->Console.log
```
I think it would be useful to access arbitrary properties on `Error.t` just like I can on `Object.t`. I also want a much easier way to work with those values and not have to switch on `Type.Classify.classify`.

First idea is to **add an `Error.get` function** just like we have `Object.get`. This would be very convenient. Other functions like `hasOwnProperty` or `keysToArray` or `copy` might make sense; I don't know.

Second idea is to **add convenience methods for classify** like this. All these do is wrap the classify function with a Switch.
```
  let classify: t => classified
  let toString: t => option<string>
  let toFloat: t => option<float>
```
The third idea is to **rename the `Type` module to `Unknown`** and maybe even have an `Unknown.t` type. Then when we have functions that return some unknown thing we could label them explicitly as `unknown` and the developer will then know they can use the `Unknown` module to classify and make sense of it. The rescript-struct module defines as `unknown` type that is the input to the parse function. TypeScript has an `unknown` type. But regardless I think `Unknown` is a more intuitive name for the module than `Type` because that is where you go to work with unknown things. You don't stay there long - just long enough to covert it to something else.

Finally and some **random thoughts on the `Type` module**.

- It won't classify a `bigint`.
- The classify symbol doesn't result in a `Symbol.t` - it is a `Type.Classify.symbol`
- The classification method uses `toString`. Why not `typeof`? Wouldn't that be more efficient?
- For a tiny module it is kind of annoying having a `Classify` sub-module just to hold the `classify` function. Couldn't this all be flattened out?

**Comment by jmagaram** [comment]
This is what I built to make it easier for me.
```rescript
module ErrorExtras = { 
    open RescriptCore
    external toObject: Error.t => {..} = "%identity"
    @get_index external get: (Error.t, string) => option<'a> = ""
    let getString: (Error.t, string) => option<string>
    let getFloat: (Error.t, string) => option<float>
    let getBool: (Error.t, string) => option<bool>
}

module Unknown = {
    external toStringUnsafe: 'a => string = "%identity"
    external toFloatUnsafe: 'a => float = "%identity"
    external toBoolUnsafe: 'a => bool = "%identity"
    external toObjectUnsafe: 'a => {..} = "%identity"
    let toObject: 'a => option<{..}>
    let toString: 'a => option<string>
    let toBool: 'a => option<bool>
    let toFloat: 'a => option<float>
}
```
And then here is how I use it to get the details I want from an exception...
```rescript
    } catch {
    | _ as exn => {
        let isUserNotFound =
          exn
          ->Error.fromException
          ->Option.flatMap(ErrorExtras.get(_, "code"))
          ->Option.flatMap(Unknown.toString)
          ->Option.filter(i => i === "auth/user-not-found")
          ->Option.isSome
```
This could be shorter, with an `Option.exists` method, directly accessing properties on an `exn`, and using the `ErrorExtras.getString`. **I think most of this should be built-in to core**, unless there is some simpler way I'm not aware of. There is hardly any real code to this stuff and much is `external`.

**Comment by jmagaram** [comment]
Some experimenting...
```rescript
  {"favorites": null}
  ->Object.get("favorites")
  ->Option.map(i => i->Array.length)
  ->Option.getWithDefault(0)
  ->Console.log // Run-time error of "Cannot read properties of null"

  {"count": null}
  ->Object.get("count")
  ->Option.map(i => i + 6)
  ->Option.getWithDefault(0)
  ->Console.log // Outputs 6, null + 6 = 6
```

**Comment by jmagaram** [comment]
I'm thinking this over. Is a JavaScript exception ALWAYS an object? If I throw a primitive does it get wrapped?

**Comment by glennsl** [comment]
No, you can throw anything in JS. Though everything in JS is an object in the OO sense.

Also, interestingly, I believe the stack trace is recorded when the `Error` object is created, not when it's thrown.

**Comment by glennsl** [comment]
> What about an `Exn.unsafeGet`? Like you proposed first. `let unsafeGet: (Exn.t, string) => option<'a>`.
> 
> Then people can decide themselves how safe they want to play it after using that.

I think the naming convention would be `getUnsafe`. Otherwise I agree :slightly_smiling_face:

**Comment by jmagaram** [comment]
This situation with getting a custom value out of `exn` or `Error` or dealing with a getter on `Object` is not ideal. (a) sometimes we don't have the getter and are considering adding `getUnsafe`. And (b) sometimes we get an `'a` which is like a TypeScript `any` but that isn't true. I can treat it as a string or array and get a runtime exception or unpredictable results.

I'd like to prototype a `Unknown` module and `unknown` type. This would be the `Type` module renamed with classify and other functions like `toString`, `toStringUnsafe`. And this could be used to pull data out of `Error` or `exn` and safely parse values retrieved from `Object.get`.

Are you interested in seeing my ideas on this?

**Comment by glennsl** [comment]
See my comment here: https://github.com/rescript-association/rescript-core/pull/108#pullrequestreview-1351445468

**Comment by cristianoc** [comment]
Btw type `unknown` exists already.
It does not have any instances.
It's currently used as the payload type of `JsError`

**Comment by woeps** [comment]
> I'm trying to do something simple. I caught a Firebase exception and want to access the `code` property that is a string. Firebase errors have code properties with contents like "auth/email-password-invalid". How can I do that? Here is one awkward way. **Is there a simpler way?**

We have a similar usecase when interacting with `aws-sdk` which occasionally may return `Error` objects with some additional properties than it's original prototype.
The pattern we currently decided to use is to type the specific error as a record and provide an identity function to "cast" any `Js.Exn.t´ to this record type.

```rescript
module CustomError = {
  type t = { message: string, code: string, otherCustomProperty: bool }
  external ofJsExn: Js.Exn.t => t = "%identity"
}

// somewhere else
switch await someAsyncStuff {
  | result => Js.Console.log(result)
  | exception Js.Exn.Error(err) =>
      let customErr = err->CustomError.ofJsExn
      customErr.code->Js.Console.log
}
```

This obviously unsafe to do. But any way to get some property of an object (in js representation) after the fact, that it was returned as an "unspecific" type would be.

The only safe way for us, I see, is to actually not use the promise api of `aws-sdk`, but the callback api and type the callback arguments with something alike `CustomError.t`. (Or catch any raised promise early and resolve it as an `result` type, so we only need to handle resolved promises down the road.)

**Comment by DZakh** [comment]
> It won't classify a bigint.

Is solved in https://github.com/rescript-association/rescript-core/pull/146

**Comment by DZakh** [comment]
> The classify symbol doesn't result in a Symbol.t - it is a Type.Classify.symbol

Is solved in https://github.com/rescript-association/rescript-core/pull/145
